### PR TITLE
cover double dependency on a prerelease (work in progress)

### LINKED ISSUE #1687: version conflicts should indicate which dependencies have conflicting requirements
When I run `pipenv install` and there's a conflict in versions, it doesn't tell me what caused those conflicting versions. This makes it unnecessarily difficult to figure out what's wrong with the dependency graph.

##### Expected result

```
Warning: your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.

- Pipfile requires requests ==2.9.1
- someotherdep requires requests >=2.18.0
- anotherdep (dependency of yetanotherdep) requires requests <=2.0
```

##### Actual result

```
Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.
  You can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.
Could not find a version that matches requests==2.9.1,>=2.18.0
Tried: 0.2.0, 0.2.1, 0.2.2, 0.2.3, 0.2.4, 0.3.0, 0.3.1, 0.3.2, 0.3.3, 0.3.4, 0.4.0, 0.4.1, 0.5.0, 0.5.1, 0.6.0, 0.6.1, 0.6.2, 0.6.3, 0.6.4, 0.6.5, 0.6.6, 0.7.0, 0.7.1, 0.7.2, 0.7.3, 0.7.4, 0.7.5, 0.7.6, 0.8.0, 0.8.1, 0.8.2, 0.8.3, 0.8.4, 0.8.5, 0.8.6, 0.8.7, 0.8.8, 0.8.9, 0.9.0, 0.9.1, 0.9.2, 0.9.3, 0.10.0, 0.10.1, 0.10.2, 0.10.3, 0.10.4, 0.10.6, 0.10.7, 0.10.8, 0.11.1, 0.11.2, 0.12.0, 0.12.1, 0.13.0, 0.13.1, 0.13.2, 0.13.3, 0.13.4, 0.13.5, 0.13.6, 0.13.7, 0.13.8, 0.13.9, 0.14.0, 0.14.1, 0.14.2, 1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.0.4, 1.1.0, 1.2.0, 1.2.1, 1.2.2, 1.2.3, 2.0.0, 2.0.0, 2.0.1, 2.0.1, 2.1.0, 2.1.0, 2.2.0, 2.2.0, 2.2.1, 2.2.1, 2.3.0, 2.3.0, 2.4.0, 2.4.0, 2.4.1, 2.4.1, 2.4.2, 2.4.2, 2.4.3, 2.4.3, 2.5.0, 2.5.0, 2.5.1, 2.5.1, 2.5.2, 2.5.2, 2.5.3, 2.5.3, 2.6.0, 2.6.0, 2.6.1, 2.6.1, 2.6.2, 2.6.2, 2.7.0, 2.7.0, 2.8.0, 2.8.0, 2.8.1, 2.8.1, 2.9.0, 2.9.0, 2.9.1,
2.9.1, 2.9.2, 2.9.2, 2.10.0, 2.10.0, 2.11.0, 2.11.0, 2.11.1, 2.11.1, 2.12.0, 2.12.0, 2.12.1, 2.12.1, 2.12.2,
2.12.2, 2.12.3, 2.12.3, 2.12.4, 2.12.4, 2.12.5, 2.12.5, 2.13.0, 2.13.0, 2.14.0, 2.14.0, 2.14.1, 2.14.1, 2.14.2, 2.14.2, 2.15.1, 2.15.1, 2.16.0, 2.16.0, 2.16.1, 2.16.1, 2.16.2, 2.16.2, 2.16.3, 2.16.3, 2.16.4, 2.16.4, 2.16.5, 2.16.5, 2.17.0, 2.17.0, 2.17.1, 2.17.1, 2.17.2, 2.17.2, 2.17.3, 2.17.3, 2.18.0, 2.18.0, 2.18.1, 2.18.1, 2.18.2, 2.18.2, 2.18.3, 2.18.3, 2.18.4, 2.18.4
```

### LINKED ISSUE #1696: --pre per–package
This is a follow up on #1677.

`pipenv lock` doesn't work because I need a preversion of a single package. `pipenv lock --pre` does work but might add other, unwanted preversion in dependencies.

I would like to propose my help in implementing this feature: if one package needs to be frozen to a preversion (and only one), allow the dependency finding to pick this one even without the `--pre` parameter.

The `--pre` parameter which the preversion search for **all** packages, and sometimes we only want one package

### PR DESCRIPTION
**Pull request by gsemet**
Work in progress. This PR adds a unit test covering the case where a dependency and a transitive dependency both want a prerelease version of the same package.

When both use the pin (`==`) operator, it works, where one use a `>=` operator, an error occurs:
```
Your dependencies could not be resolved
```
I would like to find where the two dependency are retrieved, along with their operator. I would like to add an evaluation "is the resolution not yet done for this dependency", before starting the whole "search for the best package" resolver (in pip module)

Fix #1696
Fix #1687
Reference #1693

Signed-off-by: Gaetan Semet <gaetan@xeberon.net>

### CODE REVIEW

File: tests/test_pipenv.py [hunk 1]
```diff
@@ -1064,3 +1064,41 @@ def test_install_local_file_collision(self, pypi):
...
+    def test_install_pinned_pinned_preversion(self, pypi):
+        with PipenvInstance(pypi=pypi) as p:
+            target_package = 'py2neo'
+            preversion = '==4.0.0b1'
+            transitive_pkg_name = 'neo4j-driver'
+            transitive_pkg_pinned_version = '==1.1.0rc1'
+            transitive_pkg_version = '>=1.1.0rc1'
+            c = p.pipenv('install {0}{1}'.format(target_package, preversion))
+            assert c.return_code == 0
+            c = p.pipenv('install {0}{1}'.format(transitive_pkg_name, transitive_pkg_version))
```
Comment (gsemet) [inline tests/test_pipenv.py]: this fails for the moment. I'm look for a way of making it work

**Comment by kennethreitz** [comment]
thank you for contributing actual code!

**Comment by kennethreitz** [comment]
✨🍰✨

**Comment by kennethreitz** [comment]
this stuff is super tricky — don't pull your hair out trying to figure it out :)

**Comment by kennethreitz** [comment]
If you want to hop on a Google Hangouts, let me know!

**Comment by gsemet** [comment]
greatatan@gmail.com

File: pipenv/patched/notpip/req/req_set.py [hunk 2]
```diff
@@ -455,7 +455,7 @@ def _check_skip_installed(self, req_to_install, finder):
             elif self.upgrade_strategy == "only-if-needed":
                 skip_reason = 'not upgraded as not directly required'
             else:
-                skip_reason = 'already satisfied'
+                skip_reason = 'already satisfied123'
```
Comment (kennethreitz) [inline pipenv/patched/notpip/req/req_set.py]: ?

**Comment by kennethreitz** [comment]
closing due to inactivity

**Pull request closed without merging**